src/lib/core/compatibility/compatibility.ts

```diff
@@ -0,0 +1,178 @@
+import {
+  NgModule,
+  ModuleWithProviders,
+  Directive,
+  OpaqueToken,
+  Inject,
+  Optional,
+} from '@angular/core';
+
+
+export const MATERIAL_COMPATIBILITY_MODE = new OpaqueToken('md-compatibility-mode');
+
+/** Selector that matches all elements that may have style collisions with material1. */
+export const MAT_ELEMENTS_SELECTOR = `
+  [mat-button],
+  [mat-dialog-actions],
+  [mat-dialog-close],
+  [mat-dialog-content],
+  [mat-dialog-title],
+  [mat-fab],
+  [mat-icon-button],
+  [mat-mini-fab],
+  [mat-raised-button],
+  [mat-tab-label],
+  [mat-tab-link],
+  [mat-tab-nav-bar],
+  mat-autocomplete,
+  mat-button-toggle-group,
+  mat-card,
+  mat-card-actions,
+  mat-card-content,
+  mat-card-footer,
+  mat-card-header,
+  mat-card-subtitle,
+  mat-card-title,
+  mat-card-title-group,
+  mat-checkbox,
+  mat-chip,
+  mat-dialog-actions,
+  mat-dialog-container,
+  mat-dialog-content,
+  mat-divider,
+  mat-grid-list,
+  mat-grid-tile,
+  mat-grid-tile-footer,
+  mat-grid-tile-header,
+  mat-hint,
+  mat-icon,
+  mat-ink-bar,
+  mat-input,
+  mat-list,
+  mat-list-item,
+  mat-menu,
+  mat-nav-list,
+  mat-option,
+  mat-placeholder,
+  mat-progress-bar,
+  mat-progress-circle,
+  mat-radio-button,
+  mat-radio-group,
+  mat-select,
+  mat-sidenav,
+  mat-sidenav-container,
+  mat-slider,
+  mat-spinner,
+  mat-tab,
+  mat-tab-group,
+  mat-toolbar,
+  matTooltip`;
+
+/** Selector that matches all elements that may have style collisions with material1. */
+export const MD_ELEMENTS_SELECTOR = `  
+  [md-button],
+  [md-dialog-actions],
+  [md-dialog-close],
+  [md-dialog-content],
+  [md-dialog-title],
+  [md-fab],
+  [md-icon-button],
+  [md-mini-fab],
+  [md-raised-button],
+  [md-tab-label],
+  [md-tab-link],
+  [md-tab-nav-bar],
+  md-autocomplete,
+  md-button-toggle-group,
+  md-card,
+  md-card-actions,
+  md-card-content,
+  md-card-footer,
+  md-card-header,
+  md-card-subtitle,
+  md-card-title,
+  md-card-title-group,
+  md-checkbox,
+  md-chip,
+  md-dialog-actions,
+  md-dialog-container,
+  md-dialog-content,
+  md-divider,
+  md-grid-list,
+  md-grid-tile,
+  md-grid-tile-footer,
+  md-grid-tile-header,
+  md-hint,
+  md-icon,
+  md-ink-bar,
+  md-input,
+  md-list,
+  md-list-item,
+  md-menu,
+  md-nav-list,
+  md-option,
+  md-placeholder,
+  md-progress-bar,
+  md-progress-circle,
+  md-radio-button,
+  md-radio-group,
+  md-select,
+  md-sidenav,
+  md-sidenav-container,
+  md-slider,
+  md-spinner,
+  md-tab,
+  md-tab-group,
+  md-toolbar,
+  mdTooltip`;
+
+/** Directive that enforces that the `mat-` prefix cannot be used. */
+@Directive({selector: MAT_ELEMENTS_SELECTOR})
+export class MatPrefixEnforcer {
```

Comment:
Nit: From the name, I originally thought that this would enforce that you *only* had mat- prefixes, rather than you *can't* have mat- prefixes (as in, you're enforcing that they're present). I think it's the word "Enforcer".  "MatPrefixRejector"?